### PR TITLE
feat: add kaliumPrefrences using multiplatform-settings

### PR DESCRIPTION
**Pull request by MohamadJaara**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

implementing kaliumPrefrences which wraps [multiplatform-settings](https://github.com/russhwolf/multiplatform-settings)

### Issues

preferences persistence 

### Solutions

implementing thin layer `KaliumPreferences` ontop of multiplatform-settings.
In `Android` multiplatform-settings uses [SharedPreferences](https://developer.android.com/training/data-storage/shared-preferences) which is **not** encrypted by defult but it was easy to swap it from [EncryptedSharedPreferences](https://developer.android.com/reference/androidx/security/crypto/EncryptedSharedPreferences) tht implement the same inferfaces as `SharedPreferences` but adds encryption 👍🏼.
```kotlin
actual val encryptedSettings: Settings = AndroidSettings(
        EncryptedSharedPreferences.create(
            fileName,
            masterKeyAlias,
            applicationContext,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        ), commit =  false
    )
```
In `IOS` multiplatform-settings uses `Keychain` so it supports encryption out-of-the-box.
In `JS` multiplatform-settings uses `Storage` **not** encrypted and still looking for an encrypted implemnttion of `Storage`.

As example I did rewrite the `SessionDAO` using `KaliumPreferences`.

**why KaliumPreferences ?**
my thoughts process behinde wraping `multiplatform-settings` with `KaliumPreferences` is:
1. keep the app separate from the the library (it can be replaced later with minimal refactoring needed).
2. adding new features on top of `multiplatform-settings` e.g. supporting complex data.
```kotlin
    override fun <T> putSerializable(key: String, value: T, kSerializer: KSerializer<T>) {
        encryptedSettings[key] = JsonSerializer().encodeToString(kSerializer, value)
    }

    override fun <T> getSerializable(key: String, kSerializer: KSerializer<T>): T? {
        val jsonString: String? = encryptedSettings[key]
        return jsonString?.let {
            JsonSerializer().decodeFromString(kSerializer, it)
        } ?: run { null }
```

### Testing

Unit test for `SessionDAO.kt` and `KaliumPreferences.kt`

----
#### PR Post Merge Checklist for internal contributors

 - [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Review by typfel — CHANGES_REQUESTED** [review]
Very cool! 😎

The Multiplatform-settings library look solid. Would be nice to get the tests done before merging.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/session/SessionRepository.kt [hunk 1]
```diff
@@ -1,18 +1,18 @@
...
 import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.first
 import kotlinx.coroutines.flow.flow
 import kotlinx.coroutines.flow.map
 
 interface SessionRepository {
     suspend fun storeSession(autSession: AuthSession)
-    suspend fun getSessions(): Flow<Either<CoreFailure, List<AuthSession>>>
+    suspend fun getSessions(): Either<CoreFailure, List<AuthSession>>
+    suspend fun doseSessionExist(userId: String): Either<CoreFailure, Boolean>
```
[the remark this answers is not in the document]
Comment (vitorhugods) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/session/SessionRepository.kt]: I guess we would need to serialize them, maybe (as UserId is a tuple of domain and value)? 
Or, just accept that the possibility of having two conflicting UUIDs for UserIds in the SAME device is laughable and ignore the domain, just using the value.

But in any case, I think using `UserId` in the function signature would be better, so the concern about (de)serializing to save, or just using the id stays inside the repository, not outside.

File: persistence/src/androidAndroidTest/kotlin/com/wire/kalium/presistence/client/SessionDaoImplTest.kt [hunk 2]
```diff
@@ -0,0 +1,4 @@
+package com.wire.kalium.presistence.client
+
+class SessionDaoImplTest {
```
[the remark this answers is not in the document]
Comment (MohamadJaara) [reply persistence/src/androidAndroidTest/kotlin/com/wire/kalium/presistence/client/SessionDaoImplTest.kt:3]: I was thinking about writng the `SessionDao` test in `CommonTest` and testing `KaliumPreferences` per platform

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/session/SessionScope.kt [hunk 3]
```diff
@@ -1,11 +1,11 @@
...
+import com.wire.kalium.logic.feature.auth.GetSessionsUseCase
 
 class SessionScope(
-    sessionRepository: SessionRepository
+    private val sessionRepository: SessionRepository
 ) {
-    val allSessions = GetAllSessionsUseCase(sessionRepository)
-    val saveSession = SaveSessionUseCase(sessionRepository)
+    val allSessions get() = GetSessionsUseCase(sessionRepository)
+    val saveSession get() = SaveSessionUseCase(sessionRepository)
```
Comment (gongracr) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/session/SessionScope.kt:10]: What would be the main gain of creating a new use case instance every time?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/session/SessionRepository.kt [hunk 4]
```diff
@@ -23,24 +19,33 @@ class InMemorySessionRepository : SessionRepository {
...
+    override suspend fun doesSessionExist(userId: String): Either<CoreFailure, Boolean> {
+        TODO("Not yet implemented")
+    }
 
 }
 
-class SessionRepositoryImpl(
-    private val sessionLocalDataSource: SessionLocalDataSource,
-    private val sessionMapper: SessionMapper
+class SessionDataSource(
```
Comment (gongracr) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/session/SessionRepository.kt:29]: Is there any special interest on avoiding to call this class `SessionRepositoryImpl` as other classes do?

File: persistence/src/commonMain/kotlin/com/wire/kalium/persistence/kmm_settings/KaliumPreferences.kt [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import kotlinx.serialization.decodeFromString
+import kotlinx.serialization.encodeToString
+
+class KaliumPreferences(
+    val encryptedSettings: Settings
+) {
+
+    fun remove(key: String) = encryptedSettings.remove(key)
+
+    fun exitsValue(key: String) = encryptedSettings.keys.contains(key)
```
Comment (gongracr) [inline persistence/src/commonMain/kotlin/com/wire/kalium/persistence/kmm_settings/KaliumPreferences.kt:16]: typo?
```suggestion
    fun existsValue(key: String) = encryptedSettings.keys.contains(key)
```
or maybe 
```suggestion
    fun hasValue(key: String) = encryptedSettings.keys.contains(key)
```

File: persistence/src/commonMain/kotlin/com/wire/kalium/persistence/client/SessionDao.kt [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+    override suspend fun updateCurrentSession(userId: String) = kaliumPreferences.putString(CURRENT_SESSION_KEY, userId)
+
+    override suspend fun allSessions(): DataStoreResult<Map<String, PersistenceSession>> {
+        return kaliumPreferences.getSerializable<Map<String, PersistenceSession>>(SESSIONS_KEY)?.let {
+            DataStoreResult.Success(it)
+        } ?: run { DataStoreResult.DataNotFound }
+    }
+
+    override suspend fun existSessions(): Boolean = kaliumPreferences.exitsValue(SESSIONS_KEY)
```
Comment (gongracr) [inline persistence/src/commonMain/kotlin/com/wire/kalium/persistence/client/SessionDao.kt:75]: Maybe rename to this for grammatical coherence
```suggestion
    override suspend fun sessionsExist(): Boolean = kaliumPreferences.exitsValue(SESSIONS_KEY)
```

**Comment by gongracr** [comment]
Looks great so far, just tiny comments. However, do we all agree on using a 3rd party library subject to discontinued support in the future for KMP? I'd say so in this specific case as it looks a rather straightforward one, but I still wanted to ask it.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/session/SessionRepository.kt (hunk 4 above)
Comment (MohamadJaara) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/session/SessionRepository.kt:29]: not really, just to unify the code base, i saw other devs do it so i followed

File: persistence/src/commonMain/kotlin/com/wire/kalium/persistence/kmm_settings/KaliumPreferences.kt (hunk 5 above)
Comment (MohamadJaara) [reply persistence/src/commonMain/kotlin/com/wire/kalium/persistence/kmm_settings/KaliumPreferences.kt]: changed it to `hasValue`

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/session/SessionScope.kt (hunk 3 above)
Comment (vitorhugods) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/session/SessionScope.kt:10]: @gongracr There's no need to keep a reference to one. 

If we stop using them, they can release whatever they're using internally (down to caches, persistence/storage or other IO) and free resources.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/session/SessionRepository.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (MohamadJaara) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/session/SessionRepository.kt]: we can change it to `UserId` but that will requer some some changes and i don't want to make this pr any bigger

**Review by gongracr — APPROVED** [review]
Looking good to me. However, is there a way we could test the proper encryption and decryption of the preferences even if its on the androidTest realm?

**Review by typfel — APPROVED** [review]

**Pull request merged**